src/System.IO.Pipelines/src/System/IO/Pipelines/PipeAwaitable.cs

```diff
@@ -73,39 +88,54 @@ public void Reset()
             }
         }
 
-        public bool IsCompleted => ReferenceEquals(_state, _awaitableIsCompleted);
-        internal bool HasContinuation => !ReferenceEquals(_state, _awaitableIsNotCompleted);
+        public bool IsCompleted => ReferenceEquals(_completion, _awaitableIsCompleted);
+        internal bool HasContinuation => !ReferenceEquals(_completion, _awaitableIsNotCompleted);
 
-        public Action OnCompleted(Action continuation, out bool doubleCompletion)
+        public void OnCompleted(Action<object> continuation, object state, ValueTaskSourceOnCompletedFlags flags, out Action<object> completion, out object completionState, out bool doubleCompletion)
         {
+            completionState = null;
+            completion = null;
+
             doubleCompletion = false;
-            Action awaitableState = _state;
+            Action<object> awaitableState = _completion;
             if (ReferenceEquals(awaitableState, _awaitableIsNotCompleted))
             {
-                _state = continuation;
+                _completion = continuation;
+                _completionState = state;
+
+                if ((flags & ValueTaskSourceOnCompletedFlags.UseSchedulingContext) != 0)
+                {
+                    // Set the scheduler to the current synchronization context if there is one
+                    // otherwise we delegate to what the pipe was configured with.
+                    // REVIEW: Should the sync context override the current scheduler if it was explicitly specifed
+                    // in the pipe options?
```

Comment:
> Should the sync context override the current scheduler if it was explicitly specifed in the pipe options?

Good question.  I think that depends on expected usage.  If this is expected to be a relatively closed system, where the code creating the pipe options is also the code doing all of the usage, then I think it makes sense for the pipe options to override.  If, however, unrelated code might be awaiting, e.g. if the pipe might be exposed from a library that then might be consumed on a UI thread, I think it makes sense for the current sync ctx to override.